pkg/process/monitor/process_monitor.go

```diff
@@ -119,282 +100,239 @@ type ProcessCallback struct {
+	pm.callbackRunnersWG.Add(cpuNum)
+	for i := 0; i < cpuNum; i++ {
 		go func() {
-			defer pm.wg.Done()
+			defer pm.callbackRunnersWG.Done()
 			for call := range pm.callbackRunner {
 				if call != nil {
 					call()
 				}
 			}
 		}()
 	}
+}
 
-	// This is the main async loop, where we process "processes" events from netlink socket
-	// events are dropped until
-	pm.wg.Add(1)
-	go func() {
-		logTicker := time.NewTicker(2 * time.Minute)
-
-		terminated := false
-		defer func() {
-			if !terminated {
-				pm.terminateProcessMonitor()
-			}
-			logTicker.Stop()
-		}()
+// mainEventLoop is an event loop receiving events from netlink, or periodic events, and handles them.
+func (pm *ProcessMonitor) mainEventLoop() {
+	logTicker := time.NewTicker(2 * time.Minute)
+
+	defer func() {
+		logTicker.Stop()
+		// Marking netlink to stop, so we won't get any new events.
+		close(pm.netlinkDoneChannel)
+		// waiting for the callbacks runners to finish
+		close(pm.callbackRunner)
+		// Waiting for all runners to halt.
+		pm.callbackRunnersWG.Wait()
+		// Before shutting down, making sure we're cleaning all resources.
+		pm.processMonitorWG.Done()
+	}()
 
-		for {
-			select {
-			case <-pm.done:
+	for {
+		select {
+		case <-pm.done:
+			return
+		case event, ok := <-pm.netlinkEventsChannel:
+			if !ok {
 				return
-			case event, ok := <-pm.events:
-				if !ok {
-					return
-				}
-				pm.m.Lock()
-				if !pm.isInitialized {
-					pm.m.Unlock()
-					continue
-				}
+			}
 
-				pm.eventCount += 1
-
-				switch ev := event.Msg.(type) {
-				case *netlink.ExecProcEvent:
-					for _, c := range pm.procEventCallbacks[EXEC] {
-						pm.execCount += 1
-						pm.evalEXECCallback(c, ev.ProcessPid)
-					}
-				case *netlink.ExitProcEvent:
-					for _, c := range pm.procEventCallbacks[EXIT] {
-						pm.exitCount += 1
-						pm.evalEXITCallback(c, ev.ProcessPid)
-					}
-					delete(pm.runningPids, ev.ProcessPid)
-				}
-				pm.m.Unlock()
+			pm.eventCount.Inc()
 
-			case err, ok := <-pm.errors:
-				if !ok {
-					return
-				}
-				log.Errorf("process monitor error: %s", err)
-				pm.terminateProcessMonitor()
-				terminated = true
-				pm.Stop()
+			switch ev := event.Msg.(type) {
+			case *netlink.ExecProcEvent:
+				pm.execCount.Inc()
+				pm.handleProcessExec(int(ev.ProcessPid))
+			case *netlink.ExitProcEvent:
+				pm.exitCount.Inc()
+				pm.handleProcessExit(int(ev.ProcessPid))
+			}
+		case err, ok := <-pm.netlinkErrorsChannel:
+			if !ok {
 				return
-
-			case <-logTicker.C:
-				pm.logStats()
 			}
+			log.Errorf("process monitor error: %s", err)
+			log.Info("re-initializing process monitor")
+			pm.netlinkDoneChannel <- struct{}{}
+			// Netlink might suffer from temporary errors (insufficient buffer for example). We're trying to recover
+			// by reinitializing netlink socket.
+			if err := pm.initNetlinkProcessEventMonitor(); err != nil {
+				log.Errorf("failed re-initializing process monitor: %s", err)
+				return
```

Comment:
to be on safe side to not have possible frenetic restart, could we add a waiting mechanism, in case another root process (external from datadog) do the same thing, reopening the netlink socket

kind of logic :

```
pm.retry++
wait := pm.retry*100*time.Millisecond
if wait > 3 * time.Second {
   wait = 3 * time.Second
}
time.Sleep(wait)
```

Reply:
I dont see the benefit of "extending" the time between restarts, Adding a small constant sleep

Reply:
if the another process try to steal the netlink conneciton aggressivly we will just wait 50ms and flood the logs, increasing the delay between restart would avoid higher cpu and flood logs (we will probably be throttle for that)

Reply:
seems like a very unusual case, inwhich we cannot enable https at all, and thus the customer should disable the feature...